src/main/scala/sparkz/core/network/peer/PeerManager.scala

```diff
@@ -189,34 +191,26 @@ object PeerManager {
                           sparkzContext: SparkzContext): Map[InetSocketAddress, PeerInfo] = peers.map(p => p._1 -> p._2.peerInfo)
     }
 
-    case class RandomPeerForConnectionExcluding(excludedPeers: Seq[Option[InetSocketAddress]]) extends GetPeers[Option[PeerInfo]] {
+    case class RandomPeerForConnectionExcluding(excludedPeers: Seq[Option[InetSocketAddress]], onlyKnownPeers: Boolean = false) extends GetPeers[Option[PeerInfo]] {
       private val secureRandom = new SecureRandom()
 
       override def choose(peers: Map[InetSocketAddress, PeerDatabaseValue],
                           blacklistedPeers: Seq[InetAddress],
                           sparkzContext: SparkzContext): Option[PeerInfo] = {
-        var response: Option[PeerInfo] = None
-
-        val forgerPeers = peers.filter(_._2.confidence == PeerConfidence.Forger)
-        val forgerCandidates = forgerPeers.values.filterNot(goodCandidateFilter(excludedPeers, blacklistedPeers, _)).toSeq
-
-        if (forgerCandidates.nonEmpty) {
-          response = Some(forgerCandidates(secureRandom.nextInt(forgerCandidates.size)).peerInfo)
-        } else {
-          val highConfidencePeers = peers.filter(_._2.confidence == PeerConfidence.High)
-          val highConfidenceCandidates = highConfidencePeers.values.filterNot(goodCandidateFilter(excludedPeers, blacklistedPeers, _)).toSeq
-
-          if (highConfidenceCandidates.nonEmpty) {
-            response = Some(highConfidenceCandidates(secureRandom.nextInt(highConfidenceCandidates.size)).peerInfo)
-          } else {
-            val candidates = peers.values.filterNot(goodCandidateFilter(excludedPeers, blacklistedPeers, _)).toSeq
-
-            if (candidates.nonEmpty)
-              response = Some(candidates(secureRandom.nextInt(candidates.size)).peerInfo)
-          }
-        }
-
-        response
+        val candidates: Map[PeerConfidence, Seq[PeerDatabaseValue]] = peers.values.toSeq
+          .filterNot(goodCandidateFilter(excludedPeers, blacklistedPeers, _))
+          .groupBy(_.confidence)
+
+        val forgerCandidates = candidates.getOrElse(PeerConfidence.Forger, Seq())
+        val highConfidenceCandidates = candidates.getOrElse(PeerConfidence.High, Seq())
+
+        if (forgerCandidates.nonEmpty && !onlyKnownPeers) {
+          Some(forgerCandidates(secureRandom.nextInt(forgerCandidates.size)).peerInfo)
+        } else if (highConfidenceCandidates.nonEmpty) {
```

Comment:
I think you have to add : && onlyKnownPeers
otherwise if onlyKnownPeers = false  but you have some knwow peers in the config you will always come here

Reply:
That is how it worked before, there's 2 main cases:
1. Flag is set to false, then
- first, check if there are any forgers, take random forger to connect with
- second, check if there are any high confidence peers, take random high confidence peer
- third, take random peer from any remaining peers

2. Flag is set to true, then take random peer from high confidence peers(known peers)

Am I missing something? it seems to be doing what's expected...

Reply:
I think I will decompose it into 2 several cases so that it is easier to maintain. Saving a couple lines here is not worth it...

Reply:
Maybe PeerConfidence.High could be renamed to PeerConfidence.KnownPeer, so it would be clearer? (if, of course, it is used only for known peers, as I have understood so far)

Reply:
Yes, I did that also, thanks!

Reply:
ok now is more clear!